File: vol1/ch4/LeetCode1438.cpp
class Solution {
public:
    int longestSubarray(vector<int>& nums,int limit) {
        deque<int> maxdq, mindq;
        int ans=0,n=nums.size();
        int low=0,high=0;           //´°¿ÚµÄ×óÓÒ±ß¿ò
        while (high<n) {
            while (!maxdq.empty() && maxdq.back()<nums[high])
                maxdq.pop_back();
            while (!mindq.empty() && mindq.back()>nums[high])
                mindq.pop_back();
            mindq.push_back(nums[high]);
            maxdq.push_back(nums[high]);
            while (maxdq.front()-mindq.front()>limit) {
                if (maxdq.front()==nums[low])
                    maxdq.pop_front();
                if (mindq.front()==nums[low])
                    mindq.pop_front();
                low++;
            }
             ans=max(ans,high-low+1);
             high++;
        }
        return ans;
    }
};

File: vol1/ch4/LeetCode1438.py
from collections import deque
class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        maxdq,mindq=deque(),deque()
        ans,n=0,len(nums)
        low,high=0,0           #窗口的左右边框
        while high<n:
            while len(maxdq)>0 and maxdq[-1]<nums[high]:
                maxdq.pop()
            while  len(mindq)>0 and mindq[-1]>nums[high]:
                mindq.pop()
            mindq.append(nums[high])
            maxdq.append(nums[high])
            while maxdq[0]-mindq[0]>limit:
                if maxdq[0]==nums[low]:
                    maxdq.popleft()
                if mindq[0]==nums[low]: 
                    mindq.popleft()
                low+=1
            ans=max(ans,high-low+1)
            high+=1
        return ans
